frontend/src/scenes/surveys/utils.ts

```diff
@@ -99,39 +101,107 @@ export type NPSBreakdown = {
+        processedData.data.length !== NPS_SCALE_SIZE
+    ) {
         return null
     }
 
-    if (surveyRatingResults.total === 0) {
-        return { total: 0, promoters: 0, passives: 0, detractors: 0 }
+    return {
+        values: processedData.data.map((item) => item.value),
+        total: processedData.totalResponses,
     }
+}
 
-    const PROMOTER_MIN = 9
-    const PASSIVE_MIN = 7
+/**
+ * Extracts raw NPS data from legacy survey rating results
+ */
+function extractNPSRawDataFromLegacy(surveyRatingResults: SurveyRatingResults[number]): NPSRawData | null {
+    if (!surveyRatingResults?.data || surveyRatingResults.data.length !== NPS_SCALE_SIZE) {
+        return null
+    }
 
-    const promoters = surveyRatingResults.data.slice(PROMOTER_MIN, 11).reduce((a, b) => a + b, 0)
-    const passives = surveyRatingResults.data.slice(PASSIVE_MIN, PROMOTER_MIN).reduce((a, b) => a + b, 0)
-    const detractors = surveyRatingResults.data.slice(0, PASSIVE_MIN).reduce((a, b) => a + b, 0)
-    return { total: surveyRatingResults.total, promoters, passives, detractors }
+    return {
+        values: surveyRatingResults.data,
+        total: surveyRatingResults.total,
+    }
 }
 
-export function calculateNpsScore(npsBreakdown: NPSBreakdown): number {
-    if (npsBreakdown.total === 0) {
-        return 0
+/**
+ * Core NPS calculation logic - works with raw data arrays
+ */
+function calculateNPSFromRawData(rawData: NPSRawData): NPSBreakdown {
+    if (rawData.total === 0) {
+        return { total: 0, promoters: 0, passives: 0, detractors: 0, score: '0.0' }
     }
-    return ((npsBreakdown.promoters - npsBreakdown.detractors) / npsBreakdown.total) * 100
+
+    const promoters = rawData.values.slice(NPS_PROMOTER_MIN, NPS_SCALE_SIZE).reduce((acc, curr) => acc + curr, 0)
+    const passives = rawData.values.slice(NPS_PASSIVE_MIN, NPS_PROMOTER_MIN).reduce((acc, curr) => acc + curr, 0)
+    const detractors = rawData.values.slice(0, NPS_PASSIVE_MIN).reduce((acc, curr) => acc + curr, 0)
+
+    const score = ((promoters - detractors) / rawData.total) * 100
+
+    return {
+        total: rawData.total,
+        promoters,
+        passives,
+        detractors,
+        score: score.toFixed(1),
+    }
+}
+
+export function calculateNpsBreakdownFromProcessedData(processedData: QuestionProcessedData): NPSBreakdown | null {
+    const rawData = extractNPSRawData(processedData)
+    return rawData ? calculateNPSFromRawData(rawData) : null
+}
+
+export function calculateNpsBreakdown(surveyRatingResults: SurveyRatingResults[number]): NPSBreakdown | null {
+    const rawData = extractNPSRawDataFromLegacy(surveyRatingResults)
+    return rawData ? calculateNPSFromRawData(rawData) : null
 }
 
 // Helper to escape special characters in SQL strings
 function escapeSqlString(value: string): string {
     return value.replace(/['\\]/g, '\\$&')
 }
 
+export function getSurveyResponse(question: SurveyQuestion, index: number): string {
```

Comment:
i'm recreating this function in the frontend as well (we have it on the BE) so we actually go through the HogQL parser which makes more optimized.

we have a similar one in [this file](https://github.com/PostHog/posthog/blob/feat/getting-survey-answers-in-one-query-part-1/posthog/models/surveys/util.py#L25)